### ISSUE TITLE
Loss becomes "NAN" with Mixed Precision Training

### ISSUE THREAD

**Issue by PkuRainBow** [ajabri/videowalk#11]
Really impressive work!

We have tried to support DDP and Mixed Precision Training based on your implementation while meeting a NAN issues related to: 

https://github.com/ajabri/videowalk/blob/2ac728d531acd4224177fa5fdbb8a21a7707e1af/code/model.py#L75

We have modified the above code as we used FP16:

`
A[torch.rand_like(A) < self.edgedrop_rate] = -65504`

However, we find that the loss becomes to NAN soon. It would be great if you could give me any suggestions.

**Comment by ajabri** [comment]
Hi @PkuRainBow,

Thanks for your interest! Are you sure the NaN is a result of this line; i.e. have you tried training with `--dropout 0`? 

I chose to implement dropout this way (setting the logits to a negative constant) for convenience -- you can also apply dropout on the actual transition matrices. You might also consider using a weaker negative constant -- in practice, something like `-1000` should suffice.

**Comment by PkuRainBow** [comment]
@ajabri Thanks for your suggestions and we will update the feedback to you later!

**Comment by ajabri** [comment]
Were you able to address this issue?

**Comment by ajabri** [comment]
Hi @pansanity666 

You're right that it's redundant because nn.CrossEntropy computes logsoftmax and then nll. I think I stored logprobs here because I had other losses before. You can just compute the trace of `torch.log(A+EPS)` to get the loss.

**Comment by pansanity666** [comment]
> Hi @pansanity666
> 
> You're right that it's redundant because nn.CrossEntropy computes logsoftmax and then nll. I think I stored logprobs here because I had other losses before. You can just compute the trace of `torch.log(A+EPS)` to get the loss.

Thank you for your reply. 
I tried removing the redundant  log operation:
i.e., replace
``` logits = torch.log(A+EPS).flatten(0,-2) ```
with 
``` logits = A.flatten(0,-2) ```


, and  I find it gives similar performance (or slightly better).

**Comment by ajabri** [comment]
Hi @pansanity666 

I believe `nn.CrossEntrop`y expects logits or log probabilities, so you will have to take the log.
You can do the following instead:
```
logits = torch.log(A+EPS).flatten(0,-2)
loss = -torch.einsum('bnm,nm->bn', logits, torch.eye(A.shape[-1])).mean()
``` 
```
